test/pytest/test_startup_timeout.py

```diff
@@ -0,0 +1,116 @@
+
+model_pt_file = MNIST_DIR / "mnist_cnn.pt"
+model_py_file = MNIST_DIR / "mnist.py"
+
+HANDLER_PY = """
+import time
+from ts.torch_handler.base_handler import BaseHandler
+
+class CustomHandler(BaseHandler):
+    def initialize(self, context):
+        time.sleep(10) # to simulate long startup time
+        super().initialize(context)
+
+    def handle(self, data, context):
+        return ["Dummy response"]
+"""
+
+MODEL_CONFIG_YAML = """
+minWorkers: 1
+maxWorkers: 1
+responseTimeout: 1
+startupTimeout: 120
+"""
+
+
+@pytest.fixture(scope="module")
+def model_name():
+    return "startup_timeout_test_model"
+
+
+@pytest.fixture(scope="module")
+def work_dir(tmp_path_factory):
+    return tmp_path_factory.mktemp("startup_timeout_test")
+
+
+@pytest.fixture(scope="module", name="mar_file_path")
+def create_mar_file(work_dir, model_archiver, model_name):
+    print("WORK DIR IS ", work_dir)
+    mar_file_path = work_dir / f"{model_name}.mar"
+
+    handler_file = work_dir / "handler.py"
+    handler_file.write_text(HANDLER_PY)
+
+    model_config_file = work_dir / "model_config.yaml"
+    model_config_file.write_text(MODEL_CONFIG_YAML)
+
+    config = ModelArchiverConfig(
+        model_name=model_name,
+        version="1.0",
+        serialized_file=str(model_pt_file),
+        model_file=str(model_py_file),
+        handler=str(handler_file),
+        export_path=str(work_dir),
+        config_file=str(model_config_file),
+    )
+
+    with patch("archiver.ArgParser.export_model_args_parser", return_value=config):
+        model_archiver.generate_model_archive()
+
+    assert mar_file_path.exists()
+    return str(mar_file_path)
+
+
+@pytest.fixture(scope="module")
+def register_model(mar_file_path, model_store, torchserve):
+    shutil.copy(mar_file_path, model_store)
+    model_name = Path(mar_file_path).stem
+
+    params = (
+        ("model_name", model_name),
+        ("url", Path(mar_file_path).name),
+        ("initial_workers", "1"),
+        ("synchronous", "false"),
+    )
+
+    response = test_utils.register_model_with_params(params)
+    assert response.status_code == 202, "Model registration failed"
+
+    yield model_name
+
+    test_utils.unregister_model(model_name)
+
+
+def test_startup_timeout(register_model):
+    model_name = register_model
+
+    max_wait = 30
+    start_time = time.time()
+
+    while (time.time() - start_time) < max_wait:
+        response = requests.get(f"http://localhost:8081/models/{model_name}")
+        if response.status_code == 200:
+            model_status = response.json()[0]["workers"][0]["status"]
+            if model_status == "READY":
+                break
+        time.sleep(1)
+
+    end_time = time.time()
+    elapsed_time = end_time - start_time
+
+    assert response.status_code == 200, "Model startup failed"
+    assert model_status == "READY", f"Unexpected model status: {model_status}"
+    assert 10 <= elapsed_time < 20, f"Unexpected startup time: {elapsed_time} seconds"
```

Comment:
Currently, this test only tests the successful side of registering a model. Lets also test the situation where we run into the timeout and the worker gets killed. We can just configure a 5 seconds timeout and let the init sleep 10 as its now. We can parameterize the test for this. Then we can check in the stdout of TorchServe for the right message. See [here](https://github.com/pytorch/serve/blob/3f401806742a9430f85aabba7f2fa90a7be8c56d/test/pytest/test_continuous_batching.py#L92) how to get torchserve output in a test (The tests actually ignored it but the torchserve fixture is actually a Queue that you can .get() log messaged from). Let me know if you need help with this.

Reply:
I've added the tests for it but not sure if it's the best way to do it. It reads the value from torchserve stdout and expects an error message to pop up and passes the test in that case.

Reply:
Great! That will do, thanks!